### ISSUE TITLE
Magic comment to keep manually added go_repository attributes in generated repositories

### ISSUE THREAD

**Issue by sitaktif** [bazelbuild/bazel-gazelle#895]
Some dependencies need to be amended manually to work with Gazelle. For example, the `k8s.io/api` package needs to be modified in the following way after gazelle generates the go repositories:

```python
    ...
    go_repository(
        name = "io_k8s_api",
        build_file_proto_mode = "disable_global",  # Added manually.
        importpath = "k8s.io/api",
        sum = "h1:aZ3pinA/83K1i80AH+cbrHUi/HkZ1AaToKB3q6ROZtc=",
        version = "v0.19.0-rc.1",
    )
    ...
```

I tried to put a `# keep` comment (instead of the `# Added manually` one) but gazelle will ignore it and re-generate a `go_repository` rule that does not contain the `build_file_proto_mode` attribute. I assume that's because `# keep` comments only work on non-repo rules.

It would be nice to provide a way to either ensure that the attribute gets generated by bazel, and/or a way to let the user `# keep` attributes in generated repository rules.

**Comment by jayconrod** [comment]
I couldn't reproduce this. Gazelle should add new `build_file_proto_mode` attributes to `go_repository` rules if requested, but it shouldn't change or remove those attributes later.

Please fill out the issue template. What versions of rules_go and Gazelle are you using? How are you invoking Gazelle? Where is this repository declared? `WORKSPACE` or somewhere else?

**Comment by sitaktif** [comment]
@jayconrod sorry about the slow turnaround. This is happening with recent versions of rules_go (v0.24.2) and gazelle (v0.22.0), bazel version 3.5.0. 

I initially thought it was a general issue but after I couldn't build an minimal reproducible example, I realised it only happened on this one project, which has an uncommon configuration: a root bazel WORKSPACE and an `example` sub-`WORKSPACE` which includes both its own (gazelle) go deps and the root (gazelle) go deps (I need separate bazel workspaces because what I happen to build in the root workspace is... bazel rules, so I want the example workspace to be pristine to test these rules properly. But anyway, that's besides the point, I just shared that to pre-answer the "why are you doing this?" question that was going to come :-) ).

The root workspace is rather standard, it just writes the `go_repository()` macros in `gazelle.bzl`.
The example workspace writes is own ones in `example/gazelle.bzl` but also uses `example/root-gazelle.bzl` (a symlink to the root `gazelle.bzl`) using a second `# gazelle:repository_macro` magic comment in its `WORKSPACE`.

I have a script that updates the repos with both workspaces:
1. `( cd example/ && bazelisk run //:gazelle -- update-repos -from_file=go.mod -to_macro=gazelle.bzl%go_deps_local -prune=true )`
1. `bazelisk run //:gazelle -- update-repos -from_file=go.mod -to_macro=gazelle.bzl%go_deps_local -prune=true`

What actually happens is that after step 1., the `example/root-gazelle.bzl` (which a symlink to `../gazelle.bzl`) gets pruned , removing dependencies since the example workspace doesn't use as many as the root project. Step 2 then re-adds the dependencies (and obviously the comments are not magically re-added, as that invocation would have no way to know there were there before the invocation of 1.).

What is suprising to me here is that in step 1., we only passed `gazelle.bzl` on the command line (-to_macro=gazelle.bzl%go_deps_local) but the other repo (`root-gazelle.bzl`) also got affected by the prune just by virtue of being referenced in a `# gazelle:repository_macro`.

My suggestion would be to make `-prune` only affect the macro specified on the command line.

I created https://github.com/sitaktif/gazelle-repro-895 as a minimal reproduction example (see the README to reproduce the pruning issue).

I should probably change the title of the issue to reflect the above but I'll wait to see if you'd rather I create a separate issue or if you're happy to convert this one instead.

**Comment by jayconrod** [comment]
This seems like a pretty specialized situation; I'm not really convinced `-prune` should change. If I remove a dependency, I'd like `-prune` to remove it and its dependencies from whichever macros they're declared in.

**Comment by sitaktif** [comment]
I guess either behaviour is defendable; also, I appreciate that changing behaviours has a cost. I'll give it another go, and if you're not conviced after that, so be it.

`update-repos` only mutates the file passed to `-to_macro`, while the pruning applies to all the files added via a `# gazelle:repository_macro` comment. One issue is that it is inconsistent/surprising, but the biggest issue to me is that this is no a behaviour a user would want in the first place.

Typically, files in our repository are partitioned between generated files (generally not directly modified by the user) and non-generated files (files written/modified by the user). 

When the user calls `bazelisk run //:gazelle -- update-repos -to_macro=gazelle.bzl%go_deps_local` and includes `another-gazelle.bzl%go_deps_local` repository macro in the `WORKSPACE`, it indicates that in this context:
1. the user considers `gazelle.bzl%go_deps_local` as "generated" (i.e. managed by gazelle).
1. the user considers `another-gazelle.bzl%go_deps_local` to be non-generated.

Indeed, if the user wanted dependencies in `another-gazelle.bzl%go_deps_local` to be managed by gazelle in that context, they would have put the `go_repository(...)` invocations in `gazelle.bzl%go_deps_local` in the first place.

I believe this indicates that the current behaviour does not add value compared to the one suggested above, and instead has some drawbacks (inconsistency / surprise and the impossibility to use `go_repositories()` as read-only).

**Comment by jayconrod** [comment]
> update-repos only mutates the file passed to -to_macro, while the pruning applies to all the files added via a # gazelle:repository_macro comment. One issue is that it is inconsistent/surprising, but the biggest issue to me is that this is no a behaviour a user would want in the first place.

This might be the source of confusion: `-to_macro` tells Gazelle to add new declarations to that macro, but Gazelle may still update existing declarations in `WORKSPACE` or in other macros. The expectation is that all macros declared with `# gazelle:repository_macro` are managed by Gazelle.

Gazelle won't update or delete declarations that have a `# keep` comment. I don't think that's currently supported for entire macros, but it might be a reasonable thing to add. Maybe that would solve your problem.

**Comment by sitaktif** [comment]
Ah, thanks for the clarification, that makes a lot of sense.

Allowing to `# keep` a full macro indeed solve my use case.

However, the

**Comment by jayconrod** [comment]
`# keep` comments can be applied at multiple levels of the syntax tree. A comment after an attribute only prevents that attribute from being modified, but the whole repository declaration could still be deleted.

To preserve the whole declaration, it needs to go above like this:

```
# keep
go_repository(
    name = "org_golang_x_mod",
    importpath = "golang.org/x/mod",
    sum = "h1:RM4zey1++hCTbCVQfnWeKs9/IEsaBLA8vTkd0WVtmH4=",
    version = "v0.3.0",
)
```

**Comment by sitaktif** [comment]
Thanks for the clarification, everything makes sense now.

I'll play around and if a macro-wide ends up being beneficial, I'll open a separate issue for this.

Thanks for your help in addressing these issues.
